pkg/initializr/initializr.go

```diff
@@ -0,0 +1,166 @@
+	resp, err := generate(request)
+	if err != nil {
+		return err
+	}
+	if request.Path == "" {
+		workingDir, err := os.Getwd()
+		if err != nil {
+			return err
+		}
+		return unpack(resp.Contents, workingDir)
+	}
+	return unpack(resp.Contents, request.Path)
+
+}
+
+func generate(request InitializrRequest) (InitializrResponse, error) {
+
+	u, err := url.Parse("https://start.spring.io/starter.zip")
+	if err != nil {
+		return InitializrResponse{}, err
+	}
+
+	q := u.Query()
+	q.Set("dependencies", request.Dependencies)
+	q.Set("groupId", request.GroupID)
+	q.Set("artifactId", request.ArtifactID)
+	q.Set("type", "maven-project")
+	u.RawQuery = q.Encode()
+
+	log.Debug("Initializr encoded URL: ", u.String())
+
+	// default timeout is infinite...
+	var httpClient = &http.Client{
+		Timeout: time.Second * 30,
+	}
+
+	fmt.Println("Invoking Initializr service at https://start.spring.io")
+
+	resp, err := httpClient.Get(u.String())
+	if err != nil {
+		return InitializrResponse{}, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return InitializrResponse{}, err
+	}
+
+	var initialzrResponse InitializrResponse
+	initialzrResponse.Contents = body
+	initialzrResponse.Filename = resp.Header.Get("Content-Disposition")
+	initialzrResponse.ContentType = resp.Header.Get("Content-Type")
+
+	return initialzrResponse, nil
+
+}
+
+//func closeResponse(response *http.Response) {
+//	err := response.Body.Close()
+//	if err != nil {
+//		log.Warn("Can't close http response.", err)
+//	}
+//}
+
+func unpack(zipContents []byte, targetPath string) error {
+
+	zipReader, err := zip.NewReader(bytes.NewReader(zipContents), int64(len(zipContents)))
+	if err != nil {
+		return err
+	}
+
+	// Ensure targetPath is created
+	if _, err := os.Stat(targetPath); os.IsNotExist(err) {
+		if err != nil {
+			return err
+		}
+		if err := os.MkdirAll(targetPath, os.ModePerm); err != nil {
+			return err
+		}
+	}
+
+	// Iterate through each file/dir found in
+	for _, file := range zipReader.File {
+		// Open the file inside the zip archive
+		// like a normal file
+		zippedFile, err := file.Open()
+		if err != nil {
+			return err
+		}
+		defer zippedFile.Close()
+
+		extractedFilePath := filepath.Join(
+			targetPath,
+			file.Name,
+		)
+
+		// Extract the item (or create directory)
+		if file.FileInfo().IsDir() {
+			// Create directories to recreate directory
+			// structure inside the zip archive. Also
+			// preserves permissions
+			log.Debug("Creating directory:", extractedFilePath)
+			if err := os.MkdirAll(extractedFilePath, file.Mode()); err != nil {
+				return err
+			}
+		} else {
+			// Extract regular file since not a directory
+			log.Debug("Extracting file:", file.Name)
+
+			// Open an output file for writing
+			outputFile, err := os.OpenFile(
+				extractedFilePath,
+				os.O_WRONLY|os.O_CREATE|os.O_TRUNC,
+				file.Mode(),
+			)
+			if err != nil {
+				return err
+			}
+			defer outputFile.Close()
```

Comment:
Using `defer` in a for loop is risky because the file won't be closed until all iteration finishes. It is probably a low risk as there will not be many files in the Initializr zip but worth pointing out

Reply:
Created https://github.com/cloudlift/lift/issues/14